### PR TITLE
Fix figma-json-insertion

### PR DESCRIPTION
**Pull request by vincentmvdm**
## Changes
Fixes the broken figma-json insertion by making the following changes:

- Expands write blacklist and safe assigns values that were causing errors
- Loads fonts in a way that permits failure and defaults to closest fallback font
- Inserts instances (by first importing the components)
- Uses styles (by first importing the styles)
- Adds a UI button that we can use to bulk insert figma-json test designs
    - I didn't include these designs because I don't want them in our repo, and what designs to bulk insert depends on what you're working on

See Slack for an internal video.

## Test plan
- The font code has an elaborate test suite because there's quite some things going on
- I manually tested the logic for instances/styles because it's more simple and I don't have time (yet) to write tests

## Caveats
- Doesn't yet insert images
- Doesn't yet handle overrides

## Other notes
- We should move all tests to a tests folder now that we have more of them
- We should consider moving dumping and insertion into separate files because `src/index.ts` is becoming long

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -294,27 +400,35 @@ export function decodeFont(f: EncodedFont): FontName {
   return { family, style };
 }
 
-export function preflightFonts(
```
Comment (vincentmvdm) [inline src/index.ts:297]: This code was completely unused FYI.

File: src/fonts.test.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  (globalThis as any).figma = createFigma({});
+
+  // Replace figma-api-stub's mock with one that fails to load the not installed font family.
+  const loadFontAsyncMock = jest.fn(async (font: FontName) => {
+    return new Promise<void>((resolve, reject) => {
+      if (font.family === notInstalledFontFamily) {
+        reject("Font not found");
+      }
+
+      resolve();
```
Comment (darknoon) [inline src/fonts.test.ts:26]: Looks like this should be `} else {` so as not to rely on the idea that resolve() after reject() is a no-op, right?

File: src/index.ts [hunk 3]
```diff
@@ -44,9 +44,31 @@ export const readBlacklist = new Set([
 export const writeBlacklist = new Set([
   "id",
   "componentPropertyReferences",
-  "variantProperties"
+  "variantProperties",
+  // readonly
+  "overlayPositionType",
+  "overlayBackground",
+  "overlayBackgroundInteraction",
```
Comment (darknoon) [inline src/index.ts:51]: Do these have special setters?

File: src/index.ts [hunk 4]
```diff
@@ -419,11 +533,38 @@ function applyPluginData(
...
   const offset = { x: 0, y: 0 };
   console.log("starting insert.");
 
-  await loadFonts(n);
+  // Load all the fonts, components, and styles we need in parallel.
+  const [{ fontReplacements }, { availableComponents }] = await Promise.all([
+    loadFonts(fontsToLoad(n), fallbackFonts),
+    loadComponents(n.components),
+    loadStyles(n.styles)
+  ]);
```
Comment (darknoon) [inline src/index.ts:567]: 👍

File: plugin/ui.tsx [hunk 5]
```diff
@@ -112,6 +112,19 @@ class UI extends React.Component {
     }
   };
 
+  insertTestCases = () => {
+    parent.postMessage(
+      {
+        pluginMessage: {
+          type: "insertTestCases",
+          // Add your own, local figma-json files here.
+          data: []
```
Comment (darknoon) [inline plugin/ui.tsx:121]: this comment isn't clear, could you provide more detailed directions?

**Review by darknoon — APPROVED** [review]

File: src/fonts.test.ts (hunk 2 above)
Comment (vincentmvdm) [reply src/fonts.test.ts:26]: Good call!

**Pull request merged**